x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java

```diff
@@ -0,0 +1,128 @@
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.features.NodeFeature;
+import org.elasticsearch.index.query.BoolQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.index.reindex.UpdateByQueryAction;
+import org.elasticsearch.index.reindex.UpdateByQueryRequest;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptType;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SECURITY_MAIN_ALIAS;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SecurityMainIndexMappingVersion.ADD_REMOTE_CLUSTER_AND_DESCRIPTION_FIELDS;
+
+public class SecurityMigrations {
+
+    public interface SecurityMigration {
+        /**
+         * Method that will execute the actual migration - needs to be idempotent and non-blocking
+         *
+         * @param securityIndex the main security index manager
+         * @param client the index client
+         * @param listener listener to provide updates back to caller
+         */
+        void migrate(SecurityIndexManager securityIndex, Client client, ActionListener<Void> listener);
+
+        /**
+         * Any node features that are required for this migration to run. This makes sure that all nodes in the cluster can handle any
+         * changes in behaviour introduced by the migration.
+         *
+         * @return a set of features needed to be supported or an empty set if no change in behaviour is expected
+         */
+        Set<NodeFeature> nodeFeaturesRequired();
+
+        /**
+         * The min mapping version required to support this migration. This makes sure that the index has at least the min mapping that is
+         * required to support the migration.
+         *
+         * @return the minimum mapping version required to apply this migration
+         */
+        int minMappingVersion();
+    }
+
+    public static final TreeMap<Integer, SecurityMigration> MIGRATIONS_BY_VERSION = new TreeMap<>(Map.of(1, new SecurityMigration() {
+        private static final Logger logger = LogManager.getLogger(SecurityMigration.class);
+
+        @Override
+        public void migrate(SecurityIndexManager securityIndex, Client client, ActionListener<Void> listener) {
+            BoolQueryBuilder filterQuery = new BoolQueryBuilder().filter(QueryBuilders.termQuery("type", "role"));
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder().query(filterQuery).size(0).trackTotalHits(true);
+            SearchRequest countRequest = new SearchRequest(SECURITY_MAIN_ALIAS);
+            countRequest.source(searchSourceBuilder);
+
+            client.search(countRequest, ActionListener.wrap(response -> {
```

Comment:
I figured out why/how these `client` calls, that read/write to the `.security` system index, actually work:

This is considered a "system" user action, by https://github.com/elastic/elasticsearch/blob/e77613cdfc5a2b5598ff125b5881d87f9369c867/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java#L82 
Then sytem user actions are then authz by name only: https://github.com/elastic/elasticsearch/blob/e77613cdfc5a2b5598ff125b5881d87f9369c867/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L681 
Lastly, the system user is authz for any read/write actions (on any index): https://github.com/elastic/elasticsearch/blob/e77613cdfc5a2b5598ff125b5881d87f9369c867/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/SystemPrivilege.java#L43

Reply:
TBC, I don't think switching the action origin to `SECURITY`, via `executeAsyncWithOrigin`, represents any improvement (since that user, i.e. `XPACK_SECURTITY_USER` has similar broad privileges).

Reply:
Thanks for looking in to this! I agree, let's keep it as is for now then.